File: python/batch_operations/src/occ_retry.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: MIT-0

"""OCC retry logic with exponential backoff for Aurora DSQL batch operations."""

import logging
import random
import time

import psycopg2
import psycopg2.errors

logger = logging.getLogger(__name__)


class MaxRetriesExceeded(Exception):
    """Raised when a batch operation exceeds the maximum number of OCC retries."""

    def __init__(self, max_retries):
        super().__init__(
            f"Max retries exceeded: failed after {max_retries} retries"
        )
        self.max_retries = max_retries


def execute_with_retry(connection, operation, max_retries=3, base_delay_ms=100):
    """Execute a database operation with OCC conflict retry and exponential backoff.

    Runs ``operation(connection)`` and retries on serialization failures
    (SQLSTATE 40001) using exponential backoff. Each retry rolls back the
    current transaction before waiting.

    Args:
        connection: A psycopg2 connection object.
        operation: A callable that accepts a connection and performs database
            work including commit. The operation should call ``conn.commit()``
            at the end so that commit-time serialization failures are retried.
        max_retries: Maximum number of retry attempts (default 3).
        base_delay_ms: Base delay in milliseconds for exponential backoff
            (default 100).

    Returns:
        The return value of ``operation(connection)``.

    Raises:
        MaxRetriesExceeded: If the operation fails with OCC conflicts after
            ``max_retries`` retry attempts.
    """
    for attempt in range(max_retries + 1):
        try:
            return operation(connection)
        except psycopg2.errors.SerializationFailure:
            connection.rollback()
            if attempt >= max_retries:
                raise MaxRetriesExceeded(max_retries)
            delay_ms = base_delay_ms * (2 ** attempt) * (0.5 + random.random())
            logger.warning(
                "OCC conflict, retry %d/%d after %dms",
                attempt + 1,
                max_retries,
                delay_ms,
            )
            time.sleep(delay_ms / 1000.0)

File: javascript/batch_operations/src/occRetry.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

/**
 * OCC retry logic with exponential backoff for Aurora DSQL batch operations.
 */

class MaxRetriesExceededError extends Error {
  constructor(maxRetries) {
    super(`Max retries exceeded: failed after ${maxRetries} retries`);
    this.name = "MaxRetriesExceededError";
    this.maxRetries = maxRetries;
  }
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Execute a database operation with OCC conflict retry and exponential backoff.
 *
 * Retries on SQLSTATE 40001 (serialization failure) with exponential backoff.
 * Each retry rolls back the current transaction before waiting.
 *
 * @param {import('pg').PoolClient} client - A node-postgres pool client.
 * @param {(client: import('pg').PoolClient) => Promise<*>} operation - Async
 *   function that performs database work. Must call BEGIN and COMMIT — the retry
 *   logic handles ROLLBACK on conflict before retrying.
 * @param {number} [maxRetries=3] - Maximum retry attempts.
 * @param {number} [baseDelayMs=100] - Base delay in milliseconds for backoff.
 * @returns {Promise<*>} The return value of `operation(client)`.
 */
async function executeWithRetry(client, operation, maxRetries = 3, baseDelayMs = 100) {
  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      return await operation(client);
    } catch (e) {
      if (e.code === "40001") {
        await client.query("ROLLBACK");
        if (attempt >= maxRetries) {
          throw new MaxRetriesExceededError(maxRetries);
        }
        const delayMs = baseDelayMs * Math.pow(2, attempt) * (0.5 + Math.random());
        console.warn(
          `OCC conflict, retry ${attempt + 1}/${maxRetries} after ${delayMs}ms`
        );
        await sleep(delayMs);
      } else {
        throw e;
      }
    }
  }
  throw new MaxRetriesExceededError(maxRetries);
}

module.exports = { MaxRetriesExceededError, executeWithRetry, sleep };
